### DISCUSSION TITLE
Trouble installing on Raspberry Pi 4 with 64-bit Raspberry PI OS (Debian 11 aarch64)

### DISCUSSION THREAD

**Discussion by crowellel** [qtile/qtile#3599] [Q&A] [answered]
Hello all, I've been attempting to install qtile on my Raspberry Pi 4 for the last several days and just can't get it to work!  Despite trying all manner of workarounds and suggestions that I've found by googling my problem, I keep getting the same error when I try to run the `pip install qtile` command.  Attached is the output I keep getting when I run the command.  I have been making sure to always install xcffib before cairocffi, but to no avail.  Any suggestions would be greatly appreciated!
[qtile_install_error.txt](https://github.com/qtile/qtile/files/8794398/qtile_install_error.txt)

**Comment by elParaguayo** [comment]
Sorry to hear about your problems here. I will say that I find this error to be one of the most frustrating errors when people are trying to install qtile as it seems to be a bit unpredictable.

Can you try the following:
- Uninstall both cairocffi and xcffib (`pip uninstall xcffib cairocffi`)
- Clear the pip cache (remove ~/.cache/pip, if it exists)
- Install xcffib (`pip install 'xcffib>=0.5.0'`)
- Install cairocffi (`pip install --no-deps --ignore-installed --no-cache-dir cairocffi` - yes, I know it's overkill on flags!)
- Open a python terminal and try:
```
>>> from cairocffi import cairo
>>> cairo.cairo_xcb_surface_create
```
If that last command doesn't give you an error, you should be good to go.

**Comment by crowellel** [comment]
Hi!  Thanks for your help...Unfortunately I do still get an error when I try to run the `cairo.cairo_xcb_surface_create` statement in the interpreter...I've attached a log of me following your instructions...
[error_log.txt](https://github.com/qtile/qtile/files/8799822/error_log.txt)

**Reply by elParaguayo** [reply]
So `python` defaults to version 2.7 but `pip` installs to python 3.9? That's a bit weird! Don't know if that's causing your issue or not.

Does `import xcffib` work in a python 3.9 shell session?

**Reply by crowellel** [reply]
Yeah, I thought that was odd as well...not sure how to fix that (still pretty new to linux here)...
Either way, I got no errors with `import xcffib` in the python 3.9 interpreter...

**Reply by elParaguayo** [reply]
OK.

Have a look at the last couple of comments in this thread: https://github.com/Kozea/cairocffi/issues/100

**Reply by crowellel** [reply]
So I tried the instructions at the end of that thread and once again it failed with same message.  Very frustrating!  Attached is a log of what I did..
[error_log2.txt](https://github.com/qtile/qtile/files/8801002/error_log2.txt)
.

**Reply by elParaguayo** [reply]
Frustrating!

Uninstall xcffib and cairocffi then clear the cache again.

Try the steps here: https://gitlab.com/-/snippets/2096652

Unfortunately I don't have a spare raspberry pi to try this on (and all mine are older versions too)

**Reply by crowellel** [reply]
Good Morning!

So, I had actually seen and tried that particular snippet you listed, but I went through and tried it again.  Before doing that, though, I went through and tried to clean up my python installation a little.  I found there was a debian package called 'python-is-python2' which create a symlink between the command 'python' and the python 2.7 interpreter.  I removed that and installed a different package called python-is-python3 which creates the symlink between 'python' and my current python 3.9 interpreter instead.  I also cleared out some of the old dbus-next, xcffib and cairocffi installs as well as clearing the pip3 cache.
This time after installing the dbus-next, xcffib and cairocffi[xcb] packages I again opened the interpreter and ran the
```
>>> from cairocffi import cairo
>>> cairo.cairo_xcb_surface_create
```
like you had earlier suggested.  This time it worked, so I was hopeful the install would work, but once again it failed with the same message as before!  Attached is a log with everything I did this morning
[error_log3.txt](https://github.com/qtile/qtile/files/8807347/error_log3.txt)
.

**Reply by crowellel** [reply]
So I noticed something strange and don't know what's causing this...when I run the above snippet in the interpreter in the cloned qtile directory, it seems to run fine...However, I just tried running the same snippet in the interpreter from my home directory and it gave me an error there
[error_log4.txt](https://github.com/qtile/qtile/files/8807482/error_log4.txt)
!

**Reply by elParaguayo** [reply]
I'm *really* confused now!

Can you try another line in both python shells (working and non-working). After importing cairocffi do:
```
>>> cairocffi.__file__
```
Let's see if you've got different installations.

**Reply by crowellel** [reply]
So that was something I actually checked yesterday after my previous post, and it was using separate cairocffi installations...when I noticed that, I got rid of the install that was sitting out in my home directory so all I was left with was the install I had in my ~/.local/lib/python3.9/site-packages/cairocffi/ directory.  Doing that made it so that the snippet ran without errors, but I still fail the qtile installation with the same error message as always..
[error_log5.txt](https://github.com/qtile/qtile/files/8817954/error_log5.txt)

**Comment by crowellel** [comment, marked as answer]
UPDATE:
So, after much tinkering and googling, I have finally gotten Qtile to install.  

https://gitlab.com/-/snippets/2096652

I had attempted the suggestion at the above link a couple times and it didn't work, after posting to the author of the snippet about my issues, he updated his snippet and following his updated instructions there, I was able to finally make it work.

@elParaguayo Thanks again for your time and attention!

**Reply by elParaguayo** [reply]
Fantastic. Glad you got it working.
